### PR TITLE
ISPN-6808 + ISPN-7259 Declarative config changes

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-6808
https://issues.jboss.org/browse/ISPN-7259

ISPN-6808 Validate entire cache configuration at once and aggregate exceptions
ISPN-7259 Scope-sensitive parsers for extension modules

### CODE REVIEW

**Comment by tristantarrant** [comment]
Some changes to the config validation since initial submission:
* instead of just deprecating the erroneousAttributes in CacheConfigurationException, I've decided to make them no-ops. No internal code actually used it.
* now also validates the global configuration
* wraps and tests module exceptions too

File: core/src/test/java/org/infinispan/configuration/ConfigurationUnitTest.java [hunk 1]
```diff
@@ -392,4 +398,50 @@ public void call() {
...
+         assertEquals(e.getSuppressed().length, 2);
+         assertTrue(e.getSuppressed()[0].getMessage().startsWith("ISPN000288"));
+         assertEquals("MODULE ERROR", e.getSuppressed()[1].getMessage());
+      }
+   }
+
+   public static class NonValidatingBuilder implements Builder<Object> {
+      @Override
+      public void validate() {
+         throw new CacheConfigurationException("MODULE ERROR");
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/configuration/ConfigurationUnitTest.java]: you could throw a ```RuntimeException``` to have a better simulation.

File: core/src/main/java/org/infinispan/configuration/cache/EvictionConfigurationBuilder.java [hunk 2]
```diff
@@ -101,24 +101,24 @@ public EvictionConfigurationBuilder type(EvictionType type) {
...
    public void validate() {
       EvictionStrategy strategy = attributes.attribute(STRATEGY).get();
-      Long maxEntries = attributes.attribute(SIZE).get();
+      Long size = attributes.attribute(SIZE).get();
       if (strategy == EvictionStrategy.FIFO)
          log.warnFifoStrategyIsDeprecated();
-      if (strategy.isEnabled() && maxEntries <= 0)
-         throw new CacheConfigurationException("Eviction maxEntries value cannot be less than or equal to zero if eviction is enabled");
+      if (strategy.isEnabled() && size <= 0)
+         throw log.invalidEvictionSize();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/cache/EvictionConfigurationBuilder.java]: don't forget to remove the ```CacheConfigurationException``` import :)

File: core/src/main/java/org/infinispan/configuration/parsing/Parser.java [hunk 3]
```diff
@@ -103,7 +103,7 @@
 })
 public class Parser implements ConfigurationParser {
 
-   private static final Log log = LogFactory.getLog(Parser.class);
+   static final Log log = LogFactory.getLog(Parser.class);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/parsing/Parser.java]: why is the private removed?

**Comment by tristantarrant** [comment]
Addressed all suggestions

File: core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java [hunk 4]
```diff
@@ -6,4 +6,6 @@ protected AbstractGlobalModuleConfigurationBuilder(GlobalConfigurationBuilder gl
       super(globalConfig);
    }
 
+
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java]: actually, it seems the class isn't used anymore

**Comment by pruivo** [comment]
just one small comment that I miss in the first review.

**Comment by pruivo** [comment]
The checkstyle still complains:
```
infinispan/core/src/main/java/org/infinispan/util/logging/Log.java:1470: error: Only one new line is allowed at the end of a file
```

File: core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java (hunk 4 above)
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java]: Leaving it in there to act as a base for implementors

Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java]: well, can you remove the new lines added? please?

**Comment by tristantarrant** [comment]
Everything looks good now.

File: core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java (hunk 4 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/global/AbstractGlobalModuleConfigurationBuilder.java]: is this a no to remove the new lines?

File: core/src/main/java/org/infinispan/configuration/cache/EvictionConfigurationBuilder.java (hunk 2 above)
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/configuration/cache/EvictionConfigurationBuilder.java]: That has gone :)

**Comment by tristantarrant** [comment]
Now the useless blank lines have been killed.

**Comment by pruivo** [comment]
integrated! thanks @tristantarrant !

**Pull request merged**